numpy/lib/function_base.py

```diff
@@ -1624,62 +1680,115 @@ def gradient(f, *varargs, **kwargs):
-
-            slice1[axis] = 0
-            slice2[axis] = 0
-            slice3[axis] = 1
-            slice4[axis] = 2
-            # 1D equivalent -- out[0] = -(3*y[0] - 4*y[1] + y[2]) / 2.0
-            out[slice1] = -(3.0*y[slice2] - 4.0*y[slice3] + y[slice4])/2.0
-
-            slice1[axis] = -1
-            slice2[axis] = -1
-            slice3[axis] = -2
-            slice4[axis] = -3
-            # 1D equivalent -- out[-1] = (3*y[-1] - 4*y[-2] + y[-3])
-            out[slice1] = (3.0*y[slice2] - 4.0*y[slice3] + y[slice4])/2.0
-
-        # divide by step size
-        out /= dx[i]
+            slice2[axis] = slice(None, -2)
+            slice3[axis] = slice(1, -1)
+            slice4[axis] = slice(2, None)
+            dx1 = dx[i][0:-1]
+            dx2 = dx[i][1:]
+            # Laplace polynomial coefficients
+            a = -(dx2)/(dx1 * (dx1 + dx2))
+            b = (dx2 - dx1) / (dx1 * dx2)
+            c = dx1 / (dx2 * (dx1 + dx2))
+            # fix the shape for broadcasting
+            shape = np.ones(N, dtype=int)
+            shape[axis] = -1
+            a.shape = shape
+            b.shape = shape
+            c.shape = shape
+            # 1D equivalent -- out[1:-1] = a * f[:-2] + b * f[1:-1] + c * f[2:]
+            out[slice1] = a * f[slice2] + b * f[slice3] + c * f[slice4]
+            
+            # Numerical differentiation: 1st order edges
+            if y.shape[axis] == 2 or edge_order == 1:
+                slice1[axis] = 0
+                slice2[axis] = 1
+                slice3[axis] = 0
+                # 1D equivalent -- out[0] = (y[1] - y[0]) / (x[1] - x[0])
+                out[slice1] = (y[slice2] - y[slice3]) / dx[i][0]
+                
+                slice1[axis] = -1
+                slice2[axis] = -1
+                slice3[axis] = -2
+                # 1D equivalent -- out[-1] = (y[-1] - y[-2]) / (x[-1] - x[-2])
+                out[slice1] = (y[slice2] - y[slice3]) / dx[i][-1]
+                
+                # Numerical differentiation: 2st order edges
+            else:
+                slice1[axis] = 0
+                slice2[axis] = 0
+                slice3[axis] = 1
+                slice4[axis] = 2
+                dx1 = dx[i][0]
+                dx2 = dx[i][1]
+                a = -(2. * dx1 + dx2)/(dx1 * (dx1 + dx2))
+                b = (dx1 + dx2) / (dx1 * dx2)
+                c = - dx1 / (dx2 * (dx1 + dx2))
+                # 1D equivalent -- out[0] = a * y[0] + b * y[1] + c * y[2]
+                out[slice1] = a * y[slice2] + b * y[slice3] + c * y[slice4]
+                
+                slice1[axis] = -1
+                slice2[axis] = -3
+                slice3[axis] = -2
+                slice4[axis] = -1
+                dx1 = dx[i][-2]
+                dx2 = dx[i][-1]
+                a = (dx2) / (dx1 * (dx1 + dx2))
+                b = - (dx2 + dx1) / (dx1 * dx2)
+                c = (2. * dx2 + dx1) / (dx2 * (dx1 + dx2))
```

Comment:
Done. And also optimised further...now when entering the uniform_spacing branches is ~10% faster than original np.gradient